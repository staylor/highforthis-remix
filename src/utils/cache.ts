import type { ClientLoaderFunctionArgs } from '@remix-run/react';

export const ONE_MINUTE = 60 * 1000;

interface CacheWithExpiry {
  data: any;
  expiry: number;
}

type CacheEntries = Map<string, CacheWithExpiry>;

export const createClientCache = (maxAge?: number) => {
  const expiry = maxAge || ONE_MINUTE;
  const cache: CacheEntries = new Map();
  const clientLoader = async ({ serverLoader, params }: ClientLoaderFunctionArgs) => {
    const now = Date.now();
    const key = JSON.stringify(params);
    const cached = cache.get(key);
    if (cached && now < cached.expiry) {
      return cached.data;
    }
    const data = await serverLoader();
    const entry = {} as CacheWithExpiry;
    entry.data = data;
    entry.expiry = now + expiry;
    cache.set(key, entry);
    return data;
  };
  clientLoader.hydrate = true;
  return clientLoader;
};
