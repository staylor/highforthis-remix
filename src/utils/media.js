// This file is loaded by server.js, which is not compiled.

exports.uploadUrl = (destination, fileName) => {
  return `https://storage.googleapis.com/wonderboymusic/${destination}/${fileName}`;
};
