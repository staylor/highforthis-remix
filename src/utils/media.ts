export const uploadUrl = (destination: string, fileName: string) => {
  return `https://static.highforthis.com/${destination}/${fileName}`;
};
