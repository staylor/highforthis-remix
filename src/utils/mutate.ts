import type { AppData } from '@remix-run/server-runtime';
import type { ApolloError, MutationOptions, ServerError } from '@apollo/client';
import type { GraphQLErrors } from '@apollo/client/errors';

type MutationData = Pick<MutationOptions, 'mutation' | 'variables'> & AppData;

const mutate = async ({ mutation, variables, context }: MutationData) => {
  const { apolloClient } = context;
  let data = {};
  try {
    ({ data } = await apolloClient.mutate({ mutation, variables }));
  } catch (e) {
    const error = e as ApolloError;
    console.error((error.networkError as ServerError)?.result);
    (error.graphQLErrors as GraphQLErrors).forEach((err) => {
      console.error(err.message);
    });
  }
  return data;
};

export default mutate;
