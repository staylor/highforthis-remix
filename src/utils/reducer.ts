const reducer = (a: any, b: any) => ({ ...a, ...b });

export default reducer;
