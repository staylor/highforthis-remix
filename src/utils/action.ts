import qs from 'qs';
import { parseObject } from 'query-types';
import type { AppData, DataFunctionArgs } from '@remix-run/node';
import { fetch } from '@remix-run/node';
import { redirect } from '@remix-run/node';
import type { DocumentNode } from 'graphql';
import type { OperationVariables } from '@apollo/client';

import mutate from './mutate';

export const post = async (url: string, data: AppData) =>
  fetch(url, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: { 'Content-Type': 'application/json' },
  })
    .then((response) => response.json())
    .then((responseData) => {
      if (responseData.error) {
        throw new Error(responseData.error);
      }
      return responseData;
    });

export const handleSubmission = async ({
  context,
  request,
  mutation,
  variables,
  createMutation,
}: Pick<DataFunctionArgs, 'request' | 'context'> & {
  mutation: DocumentNode;
  variables?: OperationVariables;
  createMutation?: string;
}) => {
  // FormData returns multi-dimensional keys as: foo[0][bar][baz]
  // - we would have to write our own parser, so:
  // text() returns the POST data as an x-www-form-urlencoded string
  const formData = await request.text();
  // qs parses the string into an object
  // parseObject corces the values into their proper types (numbers, booleans, etc)
  // GraphQL will throw an error if `Int`s are passed as strings.
  const input = parseObject(qs.parse(formData));
  if (input.contentState) {
    input.contentState = JSON.parse(input.contentState);
  }

  const result: AppData = await mutate({ context, mutation, variables: { ...variables, input } });
  let editUrl = request.url;
  if (createMutation) {
    editUrl = request.url.replace('/add', `/${result[createMutation].id}`);
  }

  const url = new URL(editUrl);
  url.searchParams.set('message', 'updated');

  return redirect(url.toString());
};

export const handleDelete = async ({
  request,
  context,
  mutation,
}: Pick<DataFunctionArgs, 'request' | 'context'> & { mutation: DocumentNode }) => {
  const url = new URL(request.url);
  if (request.method === 'DELETE') {
    const formData = await request.formData();
    const ids = formData.getAll('ids');

    if (ids.length > 0) {
      await mutate({
        context,
        mutation,
        variables: {
          ids,
        },
      });
    }
    url.searchParams.set('deleted', ids.length.toString());
  }
  return redirect(url.toString());
};
