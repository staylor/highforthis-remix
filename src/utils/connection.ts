import { base64Decode, base64Encode } from 'src/utils/base64';

const PREFIX = 'arrayconnection:';

export function cursorToOffset(cursor: string) {
  return parseInt(base64Decode(cursor).substring(PREFIX.length), 10);
}

export function offsetToCursor(offset: number) {
  return base64Encode(PREFIX + offset);
}
