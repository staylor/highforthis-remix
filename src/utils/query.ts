import type { AppData } from '@remix-run/server-runtime';
import type { Params } from '@remix-run/react';
import type { ApolloError, OperationVariables, QueryOptions, ServerError } from '@apollo/client';

import { PER_PAGE } from '@/constants';
import { authenticator } from '@/auth.server';

import { offsetToCursor } from './connection';

type QueryData = Pick<QueryOptions, 'query' | 'variables'> & AppData;

const query = async ({ query, variables, context, request }: QueryData) => {
  const { apolloClient } = context;
  let data = {};
  const headers: any = {};
  let authToken;
  if (request.url.includes('/admin')) {
    authToken = await authenticator.isAuthenticated(request);
  }
  if (authToken) {
    headers.Authorization = `Bearer ${authToken}`;
  }
  try {
    ({ data } = await apolloClient.query({
      query,
      variables,
      context: {
        headers,
      },
    }));
  } catch (e) {
    const error = e as ApolloError;
    if (error.graphQLErrors) {
      error.graphQLErrors.forEach((err) => {
        console.error(err.message);
      });
    }
    if (error.networkError) {
      console.error((error.networkError as ServerError)?.result?.errors);
    }
  }
  return data;
};

export default query;

export const addPageOffset = (params: Params, listVariables?: OperationVariables) => {
  const variables = listVariables || {};
  if (!variables.first) {
    variables.first = PER_PAGE;
  }
  if (params.page) {
    const pageOffset = parseInt(params.page, 10) - 1;
    if (pageOffset > 0) {
      variables.after = offsetToCursor(pageOffset * variables.first - 1);
    }
  }
  return variables;
};
