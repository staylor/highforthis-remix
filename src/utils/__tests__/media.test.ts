import { describe, expect, test } from 'vitest';
import { uploadUrl } from 'src/utils/media';

describe('media', () => {
  test('uploadUrl', () => {
    const url = uploadUrl('2018/01/31', 'foo.png');

    expect(url).toMatchSnapshot();
  });
});
