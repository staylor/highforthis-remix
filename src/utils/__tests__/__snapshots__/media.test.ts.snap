// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`media > uploadUrl 1`] = `"https://storage.googleapis.com/wonderboymusic/2018/01/31/foo.png"`;
